Keen/Query/QueryDefinition.cs

```diff
@@ -0,0 +1,118 @@
+﻿using System.Collections.Generic;
+using System;
+using System.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Keen.Core.Query
+{
+    /// <summary>
+    /// Holds information describing the query that is cached within a cached dataset.
+    /// </summary>
+    public class QueryDefinition
+    {
+        /// <summary>
+        /// Unique id of the project to analyze.
+        /// </summary>
+        public string ProjectId { get; set; }
+        /// <summary>
+        /// The type of analysis for this query (e.g. count, count_unique, sum etc.)
+        /// </summary>
+        public string AnalysisType { get; set; }
+        /// <summary>
+        /// Specifies the name of the event collection to analyze.
+        /// </summary>
+        public string EventCollection { get; set; }
+        /// <summary>
+        /// Refines the scope of events to be included in the analysis based on event property values.
+        /// </summary>
+        public IEnumerable<QueryFilter> Filters { get; set; }
+        /// <summary>
+        /// Limits analysis to a specific period of time when the events occurred.
+        /// </summary>
+        public string Timeframe { get; set; }
+        /// <summary>
+        /// Assigns a timezone offset to relative timeframes.
+        /// </summary>
+        public string Timezone { get; set; }
+        /// <summary>
+        /// Specifies the size of time interval by which to group results. Using this parameter changes the response format.
+        /// </summary>
+        public string Interval { get; set; }
+        /// <summary>
+        /// Specifies the name of a property by which to group results. Using this parameter changes the response format.
+        /// </summary>
+        public IEnumerable<string> GroupBy { get; set; }
+    }
+
+    internal static class QueryDefinitionExtensions
+    {
+        public static void ValidateForCachedDataset(this QueryDefinition query)
+        {
+            if (string.IsNullOrWhiteSpace(query.AnalysisType))
+            {
+                throw new KeenException("QueryDefinition must have an analysis type");
+            }
+
+            if (string.IsNullOrWhiteSpace(query.EventCollection))
+            {
+                throw new KeenException("QueryDefinition must specify an event collection");
+            }
+
+            if (string.IsNullOrWhiteSpace(query.Timeframe))
+            {
+                throw new KeenException("QueryDefinition must specify a timeframe");
+            }
+
+            if (string.IsNullOrWhiteSpace(query.Interval))
+            {
+                throw new KeenException("QueryDefinition must specify an interval");
+            }
+        }
+    }
+
+    /// <summary>
+    /// This is used because the PUT endpoint for a dataset take a string for group_by, but return an array of strings.
```

Comment:
I've always understood that we could pass a string or list as the `group_by` parameter, so I'm not sure why `GroupBy` being an `IEnumerable<>` doesn't just work. We'll see.

Reply:
Yeah via cURL I had no issue PUTting a new Dataset with an array for `group_by`, e.g. `"group_by" : ["price"],` so I'm fairly certain we can have an `IEnumerable<>` here.